control-plane/api/v1alpha1/servicedefaults_types.go

```diff
@@ -214,6 +218,152 @@ type ServiceDefaultsDestination struct {
+	if rl == nil {
+		return nil
+	}
+
+	return rl.InstanceLevel.validate(path.Child("instanceLevel"))
+}
+
+type InstanceLevelRateLimits struct {
+	// RequestsPerSecond is the average number of requests per second that can be
+	// made without being throttled. This field is required if RequestsMaxBurst
+	// is set. The allowed number of requests may exceed RequestsPerSecond up to
+	// the value specified in RequestsMaxBurst.
+	//
+	// Internally, this is the refill rate of the token bucket used for rate limiting.
+	RequestsPerSecond int `json:"requestsPerSecond,omitempty"`
+
+	// RequestsMaxBurst is the maximum number of requests that can be sent
+	// in a burst. Should be equal to or greater than RequestsPerSecond.
+	// If unset, defaults to RequestsPerSecond.
+	//
+	// Internally, this is the maximum size of the token bucket used for rate limiting.
+	RequestsMaxBurst int `json:"requestsMaxBurst,omitempty"`
+
+	// Routes is a list of rate limits applied to specific routes.
+	// For a given request, the first matching route will be applied, if any.
+	// Overrides any top-level configuration.
+	Routes []InstanceLevelRouteRateLimits `json:"routes,omitempty"`
+}
+
+func (irl InstanceLevelRateLimits) validate(path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Track if RequestsPerSecond is set in at least one place in the config
+	isRatelimitSet := irl.RequestsPerSecond > 0
+
+	// Top-level RequestsPerSecond can be 0 (unset) or a positive number.
+	if irl.RequestsPerSecond < 0 {
+		allErrs = append(allErrs,
+			field.Invalid(path.Child("requestsPerSecond"),
+				irl.RequestsPerSecond,
+				"RequestsPerSecond must be positive"))
+	}
+
+	if irl.RequestsPerSecond == 0 && irl.RequestsMaxBurst > 0 {
+		allErrs = append(allErrs,
+			field.Invalid(path.Child("requestsPerSecond"),
+				irl.RequestsPerSecond,
+				"RequestsPerSecond must be greater than 0 if RequestsMaxBurst is set"))
+	}
+
+	if irl.RequestsMaxBurst < 0 {
+		allErrs = append(allErrs,
+			field.Invalid(path.Child("requestsMaxBurst"),
+				irl.RequestsMaxBurst,
+				"RequestsMaxBurst must be greater than 0"))
+	}
+
+	for i, route := range irl.Routes {
+		if exact, prefix, regex := route.PathExact != "", route.PathPrefix != "", route.PathRegex != ""; (!exact && !prefix && !regex) ||
+			(exact && prefix) || (exact && regex) || (prefix && regex) {
+			allErrs = append(allErrs, field.Required(
+				path.Child("routes").Index(i),
+				"Route must define exactly one of PathExact, PathPrefix, or PathRegex"))
+		}
+
+		isRatelimitSet = isRatelimitSet || route.RequestsPerSecond > 0
+
+		// Unlike top-level RequestsPerSecond, any route MUST have a RequestsPerSecond defined.
+		if route.RequestsPerSecond <= 0 {
+			allErrs = append(allErrs, field.Invalid(
+				path.Child("routes").Index(i).Child("requestsPerSecond"),
+				route.RequestsPerSecond, "RequestsPerSecond must be greater than 0"))
+		}
+
+		if route.RequestsMaxBurst < 0 {
```

Comment:
should this be `<= 0` sine the error says `must be greater than 0`?

Reply:
if that's the case should try and test these boundaries in the unit test as well